### PR TITLE
[v2] node/relay bug fixes

### PR DESCRIPTION
**Pull request by ian-shim**
## Why are these changes needed?
A number of bug fixes in relay and v2 node.
Also, it adds a feature flag for enabling v2 features on the node. 

## Checks

- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, in that case, please comment that they are not relevant.
- [ ] I've checked the new test coverage and the coverage percentage didn't drop.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Integration tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: api/clients/relay_client.go [hunk 1]
```diff
@@ -45,49 +45,44 @@ type RelayClient interface {
...
-	grpcClients map[corev2.RelayKey]relaygrpc.RelayClient
+	// initOnce is used to ensure that the connection to each relay is initialized only once.
+	// It maps relay key to a sync.Once instance: `map[corev2.RelayKey]*sync.Once`
+	initOnce *sync.Map
+	// conns maps relay key to the gRPC connection: `map[corev2.RelayKey]*grpc.ClientConn`
+	conns  sync.Map
+	logger logging.Logger
+
+	// grpcClients maps relay key to the gRPC client: `map[corev2.RelayKey]relaygrpc.RelayClient`
+	grpcClients sync.Map
```
Comment (ian-shim) [inline api/clients/relay_client.go:56]: All maps need to be thread safe

File: relay/metadata_provider.go [hunk 2]
```diff
@@ -119,8 +127,7 @@ func (m *metadataProvider) GetMetadataForBlobs(keys []v2.BlobKey) (metadataMap,
 		}()
 	}
 
-	mMap := make(metadataMap)
-	for len(mMap) < len(keys) {
```
Comment (ian-shim) [inline relay/metadata_provider.go:123]: This hangs if there are duplicates in `keys`

Comment (jianoaix) [reply relay/metadata_provider.go:123]: Add a unit test for this

File: node/grpc/server_v2_test.go [hunk 3]
```diff
@@ -84,6 +84,23 @@ func TestV2NodeInfoRequest(t *testing.T) {
 	assert.True(t, err == nil)
 }
 
+func TestV2ServerWithoutV2(t *testing.T) {
+	config := makeConfig(t)
+	config.EnableV2 = false
+	c := newTestComponents(t, config)
+	_, err := c.server.StoreChunks(context.Background(), &pbv2.StoreChunksRequest{})
+	require.Error(t, err)
```
Comment (jianoaix) [inline node/grpc/server_v2_test.go:92]: These block of code may be benefited from a small helper function (parse error and assert the error status), as they repeat many times in this file

File: api/clients/relay_client.go [hunk 4]
```diff
@@ -45,49 +45,44 @@ type RelayClient interface {
...
 }
 
 var _ RelayClient = (*relayClient)(nil)
 
 // NewRelayClient creates a new RelayClient that connects to the relays specified in the config.
 // It keeps a connection to each relay and reuses it for subsequent requests, and the connection is lazily instantiated.
```
Comment (jianoaix) [inline api/clients/relay_client.go:62]: Why lazy init? It's not that expensive to create a connection once per Node creation

File: node/grpc/server_test.go [hunk 5]
```diff
@@ -83,6 +83,7 @@ func makeConfig(t *testing.T) *node.Config {
 		DbPath:                    t.TempDir(),
 		ID:                        opID,
 		NumBatchValidators:        runtime.GOMAXPROCS(0),
+		EnableV2:                  true,
```
Comment (jianoaix) [inline node/grpc/server_test.go:86]: This removes the test coverage for the existing v1, we need to keep testing it

File: api/clients/relay_client.go (hunk 4 above)
Comment (ian-shim) [reply api/clients/relay_client.go:62]: Just preventing a case that creates a large number of connections. This client is not necessarily only used by the node as it's meant to be used as a library

File: node/grpc/server_test.go (hunk 5 above)
Comment (ian-shim) [reply node/grpc/server_test.go]: Sounds good. Defaulted it to false and overwrote it to true in `server_v2_test`

File: node/grpc/server_v2_test.go (hunk 3 above)
Comment (ian-shim) [reply node/grpc/server_v2_test.go]: Done

File: relay/metadata_provider.go (hunk 2 above)
Comment (ian-shim) [reply relay/metadata_provider.go:123]: Done

**Review by cody-littley — APPROVED** [review]

**Pull request merged**